OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -92,54 +94,82 @@ public void Render(WorldRenderer wr)
 
-				var nextX = 0L;
-				var nextY = 0L;
-				var nextZ = 0L;
-				var k = 0;
-				for (; k < renderLength - i && k < MaxSmoothLength; k++)
+				WPos nextPos;
+				if (trailLength == 0 || renderLength == 2)
+					nextPos = curPos;
+				else
 				{
-					var prepos = trail[Index(j - k)];
-					nextX += prepos.X;
-					nextY += prepos.Y;
-					nextZ += prepos.Z;
+					// Smooth the line segment by averaging the direction of the previous segments.
+					for (var k = 1; k < renderLength - i && k < MaxSmoothLength; k++)
+					{
+						prepos = trail[Index(j - k)];
+						trailVec += prepos - pos;
+						pos = prepos;
+					}
+
+					// Normalize the vector to the length of the trail.
+					var length = (float)trailVec.Length;
+					nextPos = length == 0
+						? curPos
+						: new WPos(
+							curPos.X + (int)(trailVec.X * (float)(trailLength / length)),
+							curPos.Y + (int)(trailVec.Y * (float)(trailLength / length)),
+							curPos.Z + (int)(trailVec.Z * (float)(trailLength / length)));
 				}
 
-				var nextPos = new WPos((int)(nextX / k), (int)(nextY / k), (int)(nextZ / k));
+				var start = wr.Screen3DPosition(curPos);
+				var end = wr.Screen3DPosition(nextPos);
 
-				// When renderLength = 2 we are rendering only one segment, so it needs to be handled differently to avoid
-				// division by 0. For width we choose startWidth instead of the average as this makes the transition between
-				// rendering 1 and multiple segments smoother. Above checks make sure that renderLength can never be lower than 2.
-				float width;
-				if (renderLength == 2)
-					width = startWidth.Length;
+				// Use positions from the last segment to draw the first segment, to avoid gaps.
+				var delta = (end - start) / (end - start).XY.Length;
+				float3 startl, startr;
+				if (i == 1)
+				{
+					var startCorner = curWidth * 0.5f * new float3(-delta.Y, delta.X, delta.Z);
+					startl = start - startCorner + RgbaColorRenderer.Offset;
+					startr = start + startCorner + RgbaColorRenderer.Offset;
+				}
 				else
 				{
-					var lerp = (i - 1f) / (renderLength - 2);
-					width = startWidth.Length * (1 - lerp) + endWidth.Length * lerp;
+					startl = endl;
+					startr = endr;
 				}
 
-				if (width > 0 && !world.FogObscures(curPos) && !world.FogObscures(nextPos))
-					wcr.DrawLine(wr.Screen3DPosition(curPos), wr.Screen3DPosition(nextPos), screenWidth * width, curColor, nextColor);
+				var lerp = i / trailLengthMinusSkip;
+				var width = (startWidth.Length * (1 - lerp) + endWidth.Length * lerp) * screenWidth;
+				if (width < 0)
+					width = 0;
+
+				var endCorner = width * 0.5f * new float3(-delta.Y, delta.X, delta.Z);
+				endr = end + endCorner + RgbaColorRenderer.Offset;
+				endl = end - endCorner + RgbaColorRenderer.Offset;
+
+				var nextColor = Exts.ColorLerp(lerp, startColor, endColor);
+				if ((width != 0 || curWidth != 0)
+					&& curPos != nextPos
+					&& !world.FogObscures(curPos) && !world.FogObscures(nextPos))
```

Comment:
Won't this cause artifacts? Or are segments short enough that it is not noticable.

In theory this might draw a trail over fog if both the cur and next have no fog.